### ISSUE TITLE
Haiku output should return earlier, as soon as it has played at least one frame

### ISSUE THREAD

**Issue by MaxKellermann** [MusicPlayerDaemon/MPD#182]
The loop in `HaikuOutput::Play()` should return as soon as at least one full frame has been copied into the buffer, instead of looping until everything has been played. This allows the caller to handle commands while waiting for more buffer space.

**Comment by mmuman** [comment]
Oh, that should make a nice GCI task :-)

**Comment by sfanxiang** [comment]
Is it the same, if the caller is blocked on the next call to `HaikuOutput::Play()`, waiting for the previous operation to finish?

**Comment by MaxKellermann** [comment]
@sfanxiang I don't understand.

**Comment by sfanxiang** [comment]
@MaxKellermann 
> The loop in HaikuOutput::Play() should return as soon as at least one full frame has been copied into the buffer, instead of looping until everything has been played.

HaikuOutput::Play() is called multiple times right? So I think this is likely to happen:

Caller | Player
---------|---------
Play() |
| | fill_buffer()
Play() returns with data remaining |
(Handles other commands) |
Play() |
(Blocked) | (Plays music)
| | fill_buffer()
(Blocked) | (Plays music)
| | fill_buffer()
| | ...
| | fill_buffer() (Reaches the current buffer)
Play() returns with data remaining |
... | ...

... which only delays the loop. HaikuOutput::Play() would still have to wait until everything previously queued has been played.

**Comment by MaxKellermann** [comment]
Not quite - `Play()` would have to wait until *something* has been played, and then append more of its data to the ring buffer. So yes, `Play()` needs to block eventually if the ring buffer is completely full.
My point is that I want to reduce the amount of time `Play()` blocks. Returning from that method more often allows the caller to change his mind; let's say the user has meanwhile decided to stop playback (or to seek), then the caller may not call `Play()` again with the remaining buffer but instead discard it.

**Comment by sfanxiang** [comment]
@MaxKellermann 

But it's just a matter of time that the buffer will be filled, even before anything is actually played. For example, in one sec only some kilobytes of data can be played, while the memory can be filled with several gigabytes.

However, if the buffer is large enough to accommodate the whole song, then maybe this can reduce some blocking.

So how large would you like the buffer to be?

**Comment by MaxKellermann** [comment]
@sfanxiang I have some strange feeling you don't understand what this here is about. This issue is about an API contract between the MPD core and one of MPD's plugins. This is about implementing MPD internals. This is not about discussing or changing that API contract.
If you have ideas on how to improve MPD's internal plugin API, please open a new issue instead of hijacking this one.
But keep in mind that enlarging this buffer will certainly not reduce blocking; I think you're completely off here. It will rather put a lot of pressure on your computer, because MPD would be very busy filling this larger buffer; and eventually, it will be full, and MPD will be back to where it's now: it will block.
Blocking is **good**, because a blocking thread is idle and does not waste resources (CPU and I/O). This issue is **not** about avoiding the blocking; it is about changing *how* the thread blocks; it should be responsive, and that's the point.

**Comment by sfanxiang** [comment]
Sorry @MaxKellermann, I came from GCI and was completely unfamiliar with the code.

I did misunderstand the problem here. Is this what you want? If it's OK I'll open a PR.
``` diff
@@ -270,22 +270,12 @@ HaikuOutput::Play(const void *chunk, size_t size)
                soundPlayer->SetHasData(true);
        acquire_sem(new_buffer);
 
-       size_t bytesLeft = size;
-       while (bytesLeft > 0) {
-               if (buffer_filled == buffer_size) {
-                       // Request another buffer from BSoundPlayer
-                       release_sem(buffer_done);
-                       acquire_sem(new_buffer);
-               }
-
-               const size_t copyBytes = std::min(bytesLeft, buffer_size
-                       - buffer_filled);
-               memcpy(buffer + buffer_filled, data,
-                       copyBytes);
-               buffer_filled += copyBytes;
-               data += copyBytes;
-               bytesLeft -= copyBytes;
-       }
+       const size_t copyBytes = std::min(size, buffer_size
+               - buffer_filled);
+       memcpy(buffer + buffer_filled, data,
+               copyBytes);
+       buffer_filled += copyBytes;
+       data += copyBytes;
 
 
        if (buffer_filled < buffer_size) {
@@ -297,7 +287,7 @@ HaikuOutput::Play(const void *chunk, size_t size)
                //soundPlayer->SetHasData(false);
        }
 
-       return size;
+       return copyBytes;
 }
```

**Comment by MaxKellermann** [comment]
What is GCI?
That patch is wrong - `Play()` needs to block until at least one audio frame is copied to the buffer. It must not return `0`. And yes, please open a PR (we can discuss required changes to your code there).

**Comment by MaxKellermann** [comment]
Anyway, I really need to improve `Play()` API documentation to this clearer. I admit that too much of the API contract is only in my head .... this needs to change.

**Comment by Barrett17** [comment]
I don't understand completely why you want the function to "return more often". As long as you pass a buffer of a certain size to the function, a sane design expect that the back end try to consume the whole buffer unless something bad happened, i.e. like a locking filesystem write.

I'd say it's better you select the amount of time the buffer locks (so the buffer size) from internal code. The behavior you are requesting is not that clear and may lead to bad things. If you want the thread to consume for example half of a buffer, return, and then continue where it was at next call, then why not just pass half buffer?

I think you are trying to move a caller responsibility to the backend. If the function copied "one frame" and then returned, it would be not different than just delaying the Play() call since the sound output can play a buffer only when it's filled.

**Comment by Barrett17** [comment]
To be more simple: you control the amount of data the Play() function fill, why you want a pseudo-blocking behavior? Just control the size yourself. It won't change anything for the playback to consume "one frame" and then return, just control the size of your buffer in a different way and don't expect the backend thread to do strange things.

**Comment by sfanxiang** [comment]
@Barrett17 

> As long as you pass a buffer of a certain size to the function, a sane design expect that the back end try to consume the whole buffer unless something bad happened, i.e. like a blocking filesystem write.

The Play() function has a return value indicating the consumed size, so according to the interface the caller cannot expect the whole buffer to be consumed. Just like when you call send() and it doesn't consume the buffer, it's not exactly a bad thing.

>  If you want the thread to consume for example half of a buffer, return, and then continue where it was at next call, then why not just pass half buffer?

Because the caller doesn't know how much data can be consumed without blocking. Let's say the backend's buffer is not filled at 1/2 the caller's buffer, but at 3/4, then how can the caller do this both efficiently and non-blocking without the backend involved?

> If the function copied "one frame" and then returned, it would be not different than just delaying the Play() call since the sound output can play a buffer only when it's filled.

I didn't really follow the issue description to copy one frame in #183, but instead copied until it would block. I think it can make a difference this way, by balancing performance and responsiveness.

**Comment by Barrett17** [comment]
> The Play() function has a return value indicating the consumed size, so according to the interface the caller cannot expect the whole buffer to be consumed. Just like when you call send() and it doesn't consume the buffer, it's not exactly a bad thing.

Note this is realtime audio, not a best effort service like TCP. However, send doesn't warrant the buffer to be consumed completely, but at the same time you can't complain for it to consume the whole size passed. The situation is the same, the function just copy the whole buffer size needed by the sound card and then returns. Expecting it to return before that doesn't meet with the design I see there. 

> Because the caller doesn't know how much data can be consumed without blocking. Let's say the backend's buffer is not filled at 1/2 the caller's buffer, but at 3/4, then how can the caller do this both efficiently and non-blocking without the backend involved?

By asking the backend the exact buffer size.

But this is not very important, the caller can pass a smaller buffer size and the actual implementation won't complain. But expecting the function to return "before" is potentially harmless and inconsistent.

**Comment by MaxKellermann** [comment]
@Barrett17 sometimes, the output can consume all data at a time without blocking, and sometimes it can only copy a few bytes to the hardware. The caller cannot know that. So just saying "pass a smaller buffer" has a disadvantage (more unnecessary calls), but has no advantage. Mind that the Haiku plugin code actually gets simpler by removing the loop!

Of course it would be better to never block at all, and make the whole API non-blocking, notification based. Yes I'd very much like that, and for example the ALSA plugin is nearly ready for that. Some plugins use libraries which can never be non-blocking because of their API design. But for those who can be made non-blocking, that is a goal, and this change request is a small step in that direction.

**Comment by mmuman** [comment]
I'll have to take a closer look but IIRC in most case it shouldn't wait since the buffer is available so the semaphore can be acquired without waiting if we only send one buffer at a time.
I could add a list but it's complex enough already.

**Comment by Barrett17** [comment]
So, looking another time at the whole think I think this is what @MaxKellermann want :

* When Play() is called the code should fill one buffer, or block on it
* On the other side FillBuffer wait on a semaphore for that buffer to be filled
* When Play() fill the buffer successfully the semaphore is released, the function return the size of the buffer.
* FillBuffer consume the buffer, and so on.

@MaxKellermann correct me if I'm missing something so that @sfanxiang can do this work :)

**Comment by sfanxiang** [comment]
That's what I've done in #183 (despite the commit message). Do I need to change anything?

**Comment by MaxKellermann** [comment]
@Barrett17 sounds correct, except: you wrote that `Play()` returns the size of the buffer, and only if the semaphore is released. Both of that is wrong, but maybe this was just badly phrased. `Play()` returns the number of bytes this `Play()` invocation has consumed. And `Play()` can return without releasing the semaphore if there is not enough incoming data to fill the internal buffer completely.

@sfanxiang your commit looks correct to me, I'm just waiting for @mmuman to give green light. I'm not 100% sure if he had already agreed, because he found problems that were independent of your work.

**Comment by mmuman** [comment]
No I haven't lost interest, I just have many things on my plate.
